File: java/src/test/java/app/groups/ScenarioGroups.java
package app.groups;

import static app.config.Utils.*;
import static app.endpoints.APIendpoints.*;
import static app.endpoints.WebEndpoints.*;
import static io.gatling.javaapi.core.CoreDsl.*;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.gatling.javaapi.core.*;
import java.util.ArrayList;
import java.util.List;

public class ScenarioGroups {

  private record Product(
      int id,
      String name,
      String color,
      String price,
      int quantity,
      String imageSrc,
      String imageAlt) {}

  private static final ObjectMapper mapper = new ObjectMapper();

  private static final FeederBuilder<Object> usersFeeder =
      jsonFile("data/" + usersFeederFile).circular();

  public static final ChainBuilder homeAnonymous =
      group("homeAnonymous")
          .on(homePage, session, exec(session -> session.set("pageIndex", 0)), products);

  public static final ChainBuilder authenticate =
      group("authenticate").on(loginPage, feed(usersFeeder), pause(minSec, maxSec), login);

  public static final ChainBuilder homeAuthenticated =
      group("homeAuthenticated").on(homePage, products, pause(minSec, maxSec), search);

  public static final ChainBuilder addToCart =
      group("addToCart")
          .on(
              exec(
                  session -> {
                    try {
                      List<Product> products =
                          mapper.readValue(
                              session.getString("products"), new TypeReference<List<Product>>() {});
                      Product myFirstProduct = products.get(0);

                      List<Product> cartItems = new ArrayList<>();
                      cartItems.add(myFirstProduct);

                      String cartItemsJsonString = mapper.writeValueAsString(cartItems);
                      return session.set("cartItems", cartItemsJsonString);
                    } catch (Exception e) {
                      throw new RuntimeException(e);
                    }
                  }),
              cart);

  public static final ChainBuilder buy = group("buy").on(checkOut);
}

File: scala/src/test/scala/app/groups/ScenarioGroups.scala
package app.groups

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import io.gatling.core.structure.ChainBuilder
import com.fasterxml.jackson.core.`type`.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import java.util.ArrayList
import java.util.List
import app.config.Utils._
import app.endpoints.APIendpoints._
import app.endpoints.WebEndpoints._

object ScenarioGroups {

  case class Product(id: Int, name: String, color: String, price: Double, quantity: Int, image: String, description: String, imageSrc: String, imageAlt: String)

  val mapper = new ObjectMapper()
  mapper.registerModule(DefaultScalaModule)

  private val usersFeeder = jsonFile("data/" + usersFeederFile).circular

  val homeAnonymous: ChainBuilder = 
    group("homeAnonymous") (
      exec(homePage)
      exec(session)
      exec(session => session.set("pageIndex", 0))
      exec(products)
    )

  val authenticate: ChainBuilder = 
    group("authenticate") (
      exec(loginPage)
      feed(usersFeeder)
      pause(1, 2)
      exec(login)
    )

  val homeAuthenticated: ChainBuilder = 
    group("homeAuthenticated") (
      exec(homePage)
      exec(products)
      pause(1, 2)
      exec(search)
    )

  val addToCart: ChainBuilder = 
  group("addToCart") (
    exec { session =>
      try {
        val products: List[Product] = mapper.readValue(session("products").as[String], new TypeReference[List[Product]](){})
        val myFirstProduct: Product = products.get(0)

        val cartItems: List[Product] = new ArrayList[Product]
        cartItems.add(myFirstProduct)

        val cartItemsJsonString = mapper.writeValueAsString(cartItems)
        session.set("cartItems", cartItemsJsonString)
      } catch {
        case e: Exception => throw new RuntimeException(e)
      }
    }
    exec(cart)
  )

  val buy: ChainBuilder = 
    group("buy") (
      exec(checkOut)
    )
}
